Database manager

NichWayoe opened the pull request:
[x] Implemented suggested changes
[x] setup Database Manager

NichWayoe committed:
added contact picker

NichWayoe committed:
fix

NichWayoe committed:
made Database Manager

- moved database functions in project to one file

NichWayoe committed:
removed API key

joesus commented on FindMe/Model/Contact.m:
```diff
@@ -7,17 +7,39 @@
 
++ (void)uploadContacts: (NSArray<CNContact*> *)contacts withCompletion: (PFBooleanResultBlock  _Nullable)completion
+{
+    if(contacts) {
```
fix

joesus commented on FindMe/Model/Contact.m:
```diff
@@ -7,17 +7,39 @@
+            if(contact.givenName) {
+                newContact.lastName = contact.familyName;
+            }
+            if (contact.emailAddresses) {
```
This won't check if it's empty, or if it's the type you expect.

joesus commented on FindMe/Model/Contact.m:
```diff
@@ -7,17 +7,39 @@
+            Contact *newContact = [Contact new];
+            newContact.name = contact.givenName;
+            newContact.user = [PFUser currentUser];
+            if(contact.givenName) {
```
spacing. Also I don't understand this check. You've already used `contact.givenName` without checking for it's existence. Why here and not there?

joesus replied:
Also you're not checking for familyName.

NichWayoe replied:
Made a mistake, it was a check for family name.

joesus commented on FindMe/Model/Contact.m:
```diff
@@ -7,17 +7,39 @@
+            if (contact.emailAddresses) {
+                newContact.email = contact.emailAddresses.firstObject.value;
+            }
+            newContact.telephoneNumber = contact.phoneNumbers.firstObject.value.stringValue;
```
Could do more validation here, such as whether it's all digits or not.

joeyrhp replied:
Could this be the case if it's an international phone number with, for example, a +1 at the start?

joesus commented on FindMe/SceneDelegate.m:
```diff
@@ -7,11 +7,20 @@
+{
+    if (PFUser.currentUser) {
+          UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
+          self.window.rootViewController = [storyboard instantiateViewControllerWithIdentifier:@"Home"];}
```
Where is the else? Also, why a closing brace on the same line?

joesus commented on FindMe/Tab Bar View Contollers/MainTabBarViewController.m:
```diff
@@ -7,8 +7,14 @@
+@import ContactsUI;
 
-@interface MainTabBarViewController ()
+@interface MainTabBarViewController ()<CNContactPickerDelegate, CNContactViewControllerDelegate>
```
spacing

joesus commented on FindMe/Tab Bar View Contollers/MainTabBarViewController.m:
```diff
@@ -20,4 +26,39 @@ - (void)viewDidLoad
+            if (error != nil) {
+                NSLog(@"error %@", error.localizedDescription);
+            }
+            else {
```
Seems weird but I would also check the value of succeeded. I'm not sure if you could ever have false with no error but the method signature implies that it's possible.

joesus commented on FindMe/Tab Bar View Contollers/MainTabBarViewController.m:
```diff
@@ -20,4 +26,39 @@ - (void)viewDidLoad
+                NSLog(@"contact saved");
+            }
+        }];
+        self.selectedIndex = 2;
```
Probably want an enum for this to avoid magic numbers.

joesus commented on FindMe/Tab Bar View Contollers/MainTabBarViewController.m:
```diff
@@ -20,4 +26,39 @@ - (void)viewDidLoad
+    }
+}
+
+- (IBAction)onTapLoyout:(id)sender
```
spelling.

joesus commented on FindMe/View Controllers/ContactsViewController.h:
```diff
@@ -12,6 +12,7 @@ NS_ASSUME_NONNULL_BEGIN
 
 @interface ContactsViewController : UIViewController
 
+
```
why?

joesus commented on FindMe/View Controllers/ContactsViewController.m:
```diff
@@ -7,10 +7,14 @@
+#import "Parse/Parse.h"
 
-@interface ContactsViewController ()
+@interface ContactsViewController ()<UITableViewDataSource, UITableViewDelegate>
```
spacing

joesus commented on FindMe/View Controllers/ContactsViewController.m:
```diff
@@ -19,6 +23,43 @@ @implementation ContactsViewController
 {
     [super viewDidLoad];
+    self.contacts = [NSMutableArray new];
+    self.tableView.delegate = self;
```
can do in storyboard.

joesus commented on FindMe/View Controllers/ContactsViewController.m:
```diff
@@ -19,6 +23,43 @@ @implementation ContactsViewController
+
+- (void)fetchContacts
+{
+    PFQuery *query = [PFQuery queryWithClassName:@"Contact"];
```
VCs should not know about parse.

joesus commented on FindMe/View Controllers/ContactsViewController.m:
```diff
@@ -19,6 +23,43 @@ @implementation ContactsViewController
+- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath
+{
+    ContactCell *cell = [tableView dequeueReusableCellWithIdentifier:@"contactCell"];
+    NSLog(@"%@",self.contacts[indexPath.row]);
```
Try to not check in logging statements.

joesus commented on FindMe/View Controllers/ProfileViewController.m:
```diff
@@ -7,6 +7,10 @@
 //
 
 #import "ProfileViewController.h"
+#import "LoginViewController.h"
+#import "SceneDelegate.h"
+#import "Parse/Parse.h"
```
This is undoubtedly wrong. VCs should never need to import other VCs. Also VCs should not know about parse.

joesus commented on FindMe/Model/Contact.h:
```diff
@@ -14,12 +14,13 @@ NS_ASSUME_NONNULL_BEGIN
+@property (nonatomic, strong) NSString *email;
+@property (nonatomic, strong) NSString *name;
+@property (nonatomic, strong) NSString *lastName;
+@property (nonatomic, strong) PFUser *user;
```
Still concerns me that this is PFUser.

joesus commented on FindMe/Model/Contact.m:
```diff
@@ -32,9 +32,15 @@ + (void)uploadContacts: (NSArray<CNContact*> *)contacts withCompletion: (PFBoole
                 newContact.email = contact.emailAddresses.firstObject.value;
+            }
+            else {
+                
```
Glad to see the else's handled. What should go in here?

joesus commented on FindMe/Model/User.m:
```diff
@@ -10,15 +10,30 @@
     return @"User";
 }
 
++ (void)createUser: (NSDictionary *)userDetails  withCompletion: (PFBooleanResultBlock  _Nullable)completion{
```
This is a database operation. Should be something like `saveUser:` and be on the DatabaseManager.

joesus commented on FindMe/Model/User.m:
```diff
@@ -10,15 +10,30 @@
+    else {
+        
+    }
+    [user saveInBackgroundWithBlock: completion];}
```
This closing brace on the same line is hard to read.

joesus commented on FindMe/Parse Manager/DatabaseManager.h:
```diff
@@ -0,0 +1,22 @@
+
+@interface DatabaseManager : NSObject
+
++ (void)createUser:(NSDictionary *)userDetails withCompletion:(void(^)(NSError *error))completion;
```
This should probably take a `User` and save a `ParseUser`. Or something to that effect.

joesus commented on FindMe/Parse Manager/DatabaseManager.m:
```diff
@@ -0,0 +1,86 @@
+        user[@"profilePhoto"] = [PFFileObject fileObjectWithData:userDetails[@"profilePhotoData"]];
+    }
+    else {
+        
```
Should be something here. Should this operation fail if there's no photo?

joesus commented on FindMe/Parse Manager/DatabaseManager.m:
```diff
@@ -0,0 +1,86 @@
+    }];
+}
+
++ (void)getUser:(void(^)(PFUser * user))completion
```
Should this return a `User` or a `PFUser` ? Translating between these two is pretty key.

joesus commented on FindMe/Parse Manager/DatabaseManager.m:
```diff
@@ -0,0 +1,86 @@
+    }
+}
+
++ (void)fetchContacts:(void(^)(NSArray *contacts, bool gotContacts))completion
```
I always thought this objc pattern was sort of weird. If contacts is nil then `gotContacts` is false. This duplicates the number of cases a caller needs to handle without providing new information.

NichWayoe replied:
Yeah I tried using just one case - contacts being nil or not but the compiler kept running the unexpected resuts

joesus commented on FindMe/Parse Manager/DatabaseManager.m:
```diff
@@ -0,0 +1,86 @@
+    }];
+}
+
++ (void)logOutUser:(void(^)(bool didLogOut, NSError * error))completion
```
Same here. I think the presence of the error is all you need in order to know if the operation was successful.

joesus commented on FindMe/Parse Manager/DatabaseManager.m:
```diff
@@ -0,0 +1,86 @@
+    }];
+}
+
++ (void)logInUser:(NSString *)username withPassword:(NSString *)password withCompletion:(void(^)(bool didLogIn, NSError * error))completion
```
Same here.

joesus commented on FindMe/Tab Bar View Contollers/MainTabBarViewController.m:
```diff
@@ -42,22 +43,40 @@ - (void)contactPicker:(CNContactPickerViewController *)picker didSelectContacts:
             }
             else {
-                NSLog(@"contact saved");
+                 self.selectedIndex = 2;
```
enum would be better to avoid magic numbers.

joesus commented on FindMe/Tab Bar View Contollers/MainTabBarViewController.m:
```diff
@@ -42,22 +43,40 @@ - (void)contactPicker:(CNContactPickerViewController *)picker didSelectContacts:
-    [PFUser logOutInBackgroundWithBlock:^(NSError * _Nullable error) {
-        if (!error) {
+    [DatabaseManager logOutUser:^(bool didlogOut, NSError * _Nonnull error) {
+        if(didlogOut){
```
spacing!

joesus commented on FindMe/Tab Bar View Contollers/MainTabBarViewController.m:
```diff
@@ -42,22 +43,40 @@ - (void)contactPicker:(CNContactPickerViewController *)picker didSelectContacts:
         }
+        else {
+            [self showAlert:error];
+            
```
newline

joesus commented on FindMe/Tab Bar View Contollers/MainTabBarViewController.m:
```diff
@@ -42,22 +43,40 @@ - (void)contactPicker:(CNContactPickerViewController *)picker didSelectContacts:
+                                                            preferredStyle:(UIAlertControllerStyleAlert)];
+    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"try again"
+                                                       style:UIAlertActionStyleDefault
+                               
```
newline

joesus commented on FindMe/Tab Bar View Contollers/MainTabBarViewController.m:
```diff
@@ -42,22 +43,40 @@ - (void)contactPicker:(CNContactPickerViewController *)picker didSelectContacts:
+                                                     handler:^(UIAlertAction * _Nonnull action) {
+    }];
+    [alert addAction:okAction];
+    [self presentViewController:alert animated:YES completion:^{
```
completion:nil

joesus commented on FindMe/View Controllers/ContactsViewController.m:
```diff
@@ -62,4 +60,18 @@ - (void)viewWillAppear:(BOOL)animated
+                                                     handler:^(UIAlertAction * _Nonnull action) {
+    }];
+    [alert addAction:okAction];
+    [self presentViewController:alert animated:YES completion:^{
```
Completion: nil

joesus commented on FindMe/View Controllers/LoginViewController.m:
```diff
@@ -7,7 +7,7 @@
 
 #import "LoginViewController.h"
-#import <Parse/Parse.h>
+#import "DatabaseManager.h"
```
Way better than importing parse. (technically you are still importing parse but this makes that easier to change later)

joesus commented on FindMe/View Controllers/LoginViewController.m:
```diff
@@ -73,7 +71,7 @@ - (void)keyboardWillBeHidden:(NSNotification*)aNotification
     [UIView animateWithDuration:0.2 animations:^{self.view.frame = CGRectMake(self.view.frame.origin.x, 0, self.view.frame.size.width, self.view.frame.size.height);}];
 }
--(void) showAlert:(NSError *)error{
+-(void)showAlert:(NSError *)error{
```
👍

joesus commented on FindMe/View Controllers/SignUpViewController.m:
```diff
@@ -34,19 +34,20 @@ - (void)viewDidLoad
-    [self createBottomBoarder:self.passwordField];
-    [self createBottomBoarder:self.userNameField];
-    [self createBottomBoarder:self.confirmPasswordField];
+    [self createBottomBorder:self.firstNameField];
```
👍

joesus commented on FindMe/View Controllers/SignUpViewController.m:
```diff
@@ -91,39 +92,26 @@ - (void)createBottomBorder:(UITextField *)textField
             [self performSegueWithIdentifier:@"homeSegue" sender:nil];
-        }
-    }];
+        }}];
```
cmd-a ctrl-i

joesus commented on FindMe/View Controllers/SignUpViewController.m:
```diff
@@ -212,4 +200,19 @@ - (IBAction)onEditingConfirmPasswordField:(id)sender
+                                                     handler:^(UIAlertAction * _Nonnull action) {
+    }];
+    [alert addAction:okAction];
+    [self presentViewController:alert animated:YES completion:^{
```
completion: nil

NichWayoe committed:
fixes

- implemented user model
-moved saving of contacts to database manager

NichWayoe committed:
contactsView

- fixed a bug

NichWayoe committed:
fixed bugs

- worked on fixing consistency issues with spacing
- newline
- brackets
- indentation

NichWayoe committed:
bug fixes

NichWayoe committed:
corrected spacing
